src/wallet/wallet.cpp

```diff
@@ -4213,6 +4215,17 @@ std::shared_ptr<CWallet> CWallet::CreateWalletFromFile(interfaces::Chain& chain,
     // Try to top up keypool. No-op if the wallet is locked.
     walletInstance->TopUpKeyPool();
 
+    // Register with the validation interface. Skip requesting mempool transactions if wallet is empty.
+    interfaces::Chain::SnapshotFn snapshot_fn;
+    if (!fFirstRun) {
+        snapshot_fn = [walletInstance](std::vector<CTransactionRef> mempool_txs) {
+            for (const auto& mempool_tx : mempool_txs) {
+                walletInstance->TransactionAddedToMempool(mempool_tx);
+            }
+        };
+    }
+    walletInstance->m_chain_notifications_handler = chain.handleNotifications(*walletInstance, snapshot_fn);
```

Comment:
@jnewbery points out that it is too early to handle mempool transactions here at this point, because the rescan hasn't happened yet. Processing a mempool transaction before the rescan could result in the `IsFromMe` check in `AddToWalletIfInvolvingMe` failing to add the transaction to the walelt when it depends on a block that hasn't been scanned yet.

So the mempool loop here needs to be moved below and will require some changes to locking, since handleNotifications in this PR can't be called with cs_main (it would deadlock).

Reply:
re: https://github.com/bitcoin/bitcoin/pull/15719#discussion_r274932794

> @jnewbery points out that it is too early to handle mempool transactions here at this point

This is resolved in the current version of the PR which integrates rescans and finishes recsanning before creating the mempool snapshot